vsintegration/src/FSharp.Editor/CodeLens/LineLens.fs

```diff
@@ -0,0 +1,697 @@
+        if uiElement |> stackPanel.Children.Contains |> not then
+            self.AddUiElementToCodeLens trackingSpan uiElement
+
+    override __.RemoveUiElementFromCodeLens (trackingSpan:ITrackingSpan) (uiElement:UIElement) =
+        let stackPanel = uiElements.[trackingSpan]
+        stackPanel.Children.Remove(uiElement) |> ignore
+
+type internal FSharpCodeLensService
+    (
+        workspace: Workspace, 
+        documentId: Lazy<DocumentId>,
+        buffer: ITextBuffer, 
+        checker: FSharpChecker,
+        projectInfoManager: FSharpProjectOptionsManager,
+        typeMap: Lazy<ClassificationTypeMap>,
+        gotoDefinitionService: FSharpGoToDefinitionService,
+        codeLens : CodeLensDisplayService
+     ) as self =
+
+    let lineLens = codeLens
+
+    let visit pos parseTree = 
+        AstTraversal.Traverse(pos, parseTree, { new AstTraversal.AstVisitorBase<_>() with 
+            member this.VisitExpr(_path, traverseSynExpr, defaultTraverse, expr) =
+                defaultTraverse(expr)
+            
+            override this.VisitInheritSynMemberDefn (_, _, _, _, range) = Some range
+
+            override this.VisitTypeAbbrev( _, range) = Some range
+
+            override this.VisitLetOrUse(binding, range) = Some range
+
+            override this.VisitBinding (fn, binding) =
+                Some binding.RangeOfBindingAndRhs
+        })
+
+    let formatMap = lazy typeMap.Value.ClassificationFormatMapService.GetClassificationFormatMap "tooltip"
+
+    let mutable lastResults = Dictionary<string, ITrackingSpan * CodeLens>()
+    let mutable firstTimeChecked = false
+    let mutable bufferChangedCts = new CancellationTokenSource()
+    let uiContext = SynchronizationContext.Current
+
+    let layoutTagToFormatting (layoutTag: LayoutTag) =
+        layoutTag
+        |> RoslynHelpers.roslynTag
+        |> ClassificationTags.GetClassificationTypeName
+        |> typeMap.Value.GetClassificationType
+        |> formatMap.Value.GetTextProperties   
+
+    let createTextBox (lens:CodeLens) =
+        asyncMaybe {
+            let! taggedText, navigation = lens.TaggedText
+            let textBox = new TextBlock(Width = 500., Background = Brushes.Transparent, Opacity = 0.1, TextTrimming = TextTrimming.None)
+            DependencyObjectExtensions.SetDefaultTextProperties(textBox, formatMap.Value)
+            textBox.Inlines.Add (Documents.Run Settings.CodeLens.Prefix)
+            for text in taggedText do
+                let run = Documents.Run text.Text
+                DependencyObjectExtensions.SetTextProperties (run, layoutTagToFormatting text.Tag)
+                let inl =
+                    match text with
+                    | :? Layout.NavigableTaggedText as nav when navigation.IsTargetValid nav.Range ->
+                        let h = Documents.Hyperlink(run, ToolTip = nav.Range.FileName)
+                        h.Click.Add (fun _ -> navigation.NavigateTo nav.Range)
```

Comment:
This does only work as long as the original range is valid in terms of the same text is still there. Making changes to the file violates this which results in wrong navigation. 

I suggest either to track it or to use the goto definition service. Last one is preferred. There are already enough complete user implementations of features in this PR.